proposal for test folder consolidation (#191)
kyma-project/community: sig-and-wg/sig-core/proposals/test-folder-consolidation.md

# Test folder consolidation

## Introduction

> The [tests](https://github.com/kyma-project/kyma/tree/master/tests) folder of the kyma-project contains tests executed against a running Kyma cluster to assure integrity and functional correctness of the cluster with all modules installed.

Reviewer: ```suggestion
The [acceptance tests](https://github.com/kyma-project/kyma/tree/master/tests) folder of the kyma-project contains tests executed against a running Kyma cluster to assure integrity and functional correctness of the cluster with all modules installed.
```
I'm wondering if we maybe can merge the line saying `these are acceptance tests` - in the link name?

Author: The folder is called "tests", not acceptance tests

Reviewer: Ok - I get it. Let's leave it as it is.

> The [tests](https://github.com/kyma-project/kyma/tree/master/tests) folder of the kyma-project contains all tests executed against a running kyma cluster to assure integrity and functional correctness of the cluster when all modules are installed together.

Reviewer: ```suggestion
The [tests](https://github.com/kyma-project/kyma/tree/master/tests) folder of the kyma-project contains tests executed against a running Kyma cluster to assure integrity and functional correctness of the cluster with all modules installed.
```

> These are acceptance tests.

Reviewer: ```suggestion

```

Remove the line and merge it with the paragraph above.

Author: No, I don't like the idea. We are talking about the tests contained in that folder. I want to express that these tests are covering one test suite each and they act in the role of acceptance tests. That consideration should be the outcome of the paragraph for the reader.

Reviewer: Ok, sure - let's leave it then.

> In kyma, that kind of tests are called _acceptance tests_.

Reviewer: ```suggestion
These are acceptance tests.
```

> All subfolders in the _tests_ directory define one test suite, usually focussing on one component.

Reviewer: ```suggestion
All subfolders in the `tests` directory belong to one test suite, usually focusing on one component.
```

Author: One folder means one test suite, so they are really defining it
```suggestion
All subfolders in the `tests` directory define one test suite, usually focusing on one component.
```

## Problem

> Each subfolder uses a different naming style. For example, `kubeless` or `ui-api-layer-acceptance-tests`.

Reviewer: ```suggestion
Each subfolder uses a different naming pattern. For example, `kubeless` or `ui-api-layer-acceptance-tests`.
```

> Every subfolder uses a different naming style, like _kubeless_ and _ui-api-layer-acceptance-tests_.

Reviewer: ```suggestion
Each subfolder uses a different naming style. For example, `kubeless` or `ui-api-layer-acceptance-tests`.
```

> With that:

Reviewer: ```suggestion
As a result:
```

> - Overall it is not easy to decide which convention to follow when creating a new subfolder.

Reviewer: ```suggestion
- It's difficult to decide which convention to follow when creating a new subfolder.
```

> - It is not always possible to derive the related component (there is no e2e test yet, the tests are all focused on components).

Reviewer: ```suggestion
- It's not always possible to derive the related component (there is no e2e test yet, the tests are focused on components).
```

> - It is confusing for readers.

Reviewer: ```suggestion
- It's confusing for readers.
```

> - Not always it is possible to derive the related component (there is no e2e test yet, they are all focussed on components).

Reviewer: ```suggestion
- It is not always possible to derive the related component (there is no e2e test yet, the tests are all focused on components).
```

> - It is confusing for readers

Reviewer: ```suggestion
- It is confusing for readers.
```

## Goal

> Propose a name pattern which places the valuable information, the differentiator of the folder, into the focus.

Reviewer: ```suggestion
Propose a name pattern using specific information to distinguish between the folders.
```

## Consideration

> **kind of test** All tests are of the same style currently, named as _acceptance tests_. As long as there is no different kind of tests available, the information of the kind of test is not valuable as part of the subfolder name

Reviewer: ```suggestion
**kind of test** Currently, all tests are called **acceptance tests**. As long as there is no different kind of tests available, the test kind is not valuable information as part of the subfolder name.
```

> **test prefix/suffix** all subfolders contain test suites. That's why they are located under the _tests_ folder. Having a prefix or suffix in the folder name is redundant and does not bring more value to the reader.

Reviewer: ```suggestion
**test prefix/suffix** All subfolders contain test suites. That's why they reside in the `test` folder. A prefix or suffix in the folder name is redundant and does not bring more value to the reader.
```

> **one component per folder** as long as a test suite is focussing on a specific component under test, the test suite should be named accordingly. With that the component and also the maintainers can be identified more quickly.

Reviewer: ```suggestion
**one component per folder** As long as a test suite focuses on a specific component, the test suite should be named accordingly. This way the component and its maintainers can be identified quicker.
```

> There is one special folder called _acceptance_ currently containing test suites for 3 different components, but sharing some few common scripts. As the tendency is to have a dedicated folder and docker image per component, that folder should be converted into three dedicated component folders with own docker images. Common scripts can be still shared on the root level or a dedicated _common_ folder.

Reviewer: ```suggestion
The `acceptance` folder currently contains test suites for 3 different components. These suites share some few common scripts. The best practice is to have a dedicated folder per component, including a docker image. That's why the `acceptance` folder should be converted into three dedicated component folders with their own docker images. Common scripts can be still shared on the root level or in a `common` folder.
```

> **docker images** Every subfolder is producing a docker image using the folder name as image name. If no prefix/suffix is used, the name will collide with the image of the actual component. To have a differentiation here, a subfolder for the image could be introduced

Reviewer: ```suggestion
**docker images** Each subfolder produces a docker image using the folder name for the image name. If no prefix/suffix is used, the folder and image names can be conflicting. Introducing a subfolder for the image could solve this problem. 
```

## Proposal

Use the component's name for the component-specific subfolder.

No prefix, no suffix, no `acceptance` no multi-components.

> The docker image for a subfolder should reside in the `tests` subfolder.

Reviewer: ```suggestion
The docker image for a subfolder resides in the `tests` subfolder.
```

> The subfolder specific to a component should be named like the component, that's it.

Reviewer: ```suggestion
Use the component's name for the component-specific subfolder.
```

> No prefix, no suffix, no _acceptance_, no multi-components

Reviewer: ```suggestion
No prefix, no suffix, no `acceptance` no multi-components.
```

> **Example**: The Event-Bus component has its acceptance tests in the `tests/event-bus` folder and produces a docker image named like `XX/tests/event-bus:0.5.1`

Reviewer: ```suggestion
**Example**: The Event-Bus component has its acceptance tests in the `tests/event-bus` folder and produces the `XX/tests/event-bus:0.5.1` docker image.
```

> The resulting docker image of a subfolder should be located in the subfolder _tests_

Reviewer: ```suggestion
The docker image for a subfolder should reside in the `tests` subfolder.
```

> **Example**: A component _event-bus_ has its acceptance tests in folder _tests/event-bus_ and produces a docker image _XX/tests/event-bus:0.5.1_

Reviewer: ```suggestion
**Example**: The Event-Bus component has its acceptance tests in the `tests/event-bus` folder and produces a docker image located in `XX/tests/event-bus:0.5.1`
```

Author: ```suggestion
**Example**: The Event-Bus component has its acceptance tests in the `tests/event-bus` folder and produces a docker image named like `XX/tests/event-bus:0.5.1`
```

> + Bundle the real e2e scenarios (like kubeless-integration) into one `e2e` subfolder. Here we should have one test project which executes all e2e tests divided by scenarios to different packages.

Reviewer: ```suggestion
+ Bundle the real e2e scenarios (like kubeless-integration) into one `e2e` subfolder. Here we should have one test project which executes all e2e tests divided by scenarios into different packages.
```

## Actions

- Include a `README.md` file in the `tests` folder to explain the naming convention.

> - Link the `README.md` file in the developer guide.

Reviewer: ```suggestion
- Link the `README.md` file in the developer's guide.
```

> Real e2e scenarios (like kubeles--integration) should be bundled into one subfolder _e2e_. Here we should have one test project which will execute all e2e tests organized by scenarios in different packages.

Reviewer: ```suggestion
Bundle the real e2e scenarios (like Kubeless--integration) into one `e2e` subfolder. Here we should have one test project which executes all e2e tests divided by scenarios to different packages.
```

Author: ```suggestion
+ Bundle the real e2e scenarios (like kubeless-integration) into one `e2e` subfolder. Here we should have one test project which executes all e2e tests divided by scenarios to different packages.
```

> ## Concrete Actions

Reviewer: ```suggestion
## Actions
```
Or Action Points

> - Migrate the `kubeless-integration_ to _e2e` project containing e2e scenarios on the package level to have one test suite.

Reviewer: ```suggestion
- Migrate the `kubeless-integration` to `e2e` project containing e2e scenarios on the package level to have one test suite.
```

> - Have a readme in the _tests_ folder explaining the naming convention

Reviewer: ```suggestion
- Include a `README.md` file in the `tests` folder to explain the naming convention.
```

> - Link the readme in the developer guide

Reviewer: ```suggestion
- Link the `README.md` file in the developer guide.
```

> - Migrate _kubeless-integration_ to _e2e_ project containing e2e scenarios on a per package level resulting in one test suite to be executed.

Reviewer: ```suggestion
- Migrate the `kubeless-integration_ to _e2e` project containing e2e scenarios on the package level to have one test suite.
```

> - All folders will stay in the root _tests_ folder but will be renamed according to the following table:

Reviewer: ```suggestion
- Keep all folders in the root `tests` folder rename them according to the information in this table:
```

| old folder | new folder | action required |
|------------|------------|-----------------|
| acceptance | dex | yes |
| _ | service-catalog | yes |

> | _ | application-controller | yes |

Author: So you recommend to move it into the proposed e2e test suite, or should it be splitted somehow?
